### PR TITLE
Circular validation cleanups, part 5

### PR DESCRIPTION
**Pull request by slavapestov**
Follow-up to https://github.com/apple/swift/pull/27553.

First, this PR refactors interface type computation for decls with parameter lists (functions, constructors, subscripts, enum elements) so that instead of `validateDecl()` setting the interface type of all `ParamDecls` as a side-effect of validating the parent decl, we now end up calling `validateDecl()` on each individual `ParamDecl`. This completely eliminates `typeCheckParameterList()` and `validateParameterType()`.

Second, the logic that was moved from `validateParameterType()` to `validateDecl()` also set the `ParamDecl`'s specifier as a side effect. Factor this out into its own request.

### CODE REVIEW

**Comment by slavapestov** [comment]
@xedin This PR is blocked on https://github.com/apple/swift/pull/26973, because compiler_crashers_fixed/28723-unreachable-executed-at-swift-lib-sema-csdiag-cpp-4012.swift fails.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by theblixguy** [comment]
@hamishknight is working on a new fix for SR-11085 which should resolve that issue.

**Comment by slavapestov** [comment]
@theblixguy @hamishknight That's excellent! It wouldn't be the end of the world if we had to XFAIL that test also, but getting a clean fix in is obviously better. :)

File: include/swift/AST/Decl.h [hunk 1]
```diff
@@ -5158,6 +5161,13 @@ class VarDecl : public AbstractStorageDecl {
   }
 };
 
+enum class ParamSpecifier : uint8_t {
```
Comment (CodaFi) [inline include/swift/AST/Decl.h]: 'long as we're here, can we pick a better name than "Specifier"?

Comment (slavapestov) [reply include/swift/AST/Decl.h]: Sure, I'm all ears :) I believe one of your refactorings introduced the name Specifier, but I might be wrong :)

Comment (CodaFi) [reply include/swift/AST/Decl.h]: I did it back then because the it's the name the Swift Book's "grammar" gives it.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please smoke test Linux

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility Release

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test macOS

**Comment by slavapestov** [comment]
@swift-ci Please smoke test macOS

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Pull request merged**